services/object-mapping-indexer/src/services/objectMappingRouter/index.ts

```diff
@@ -106,25 +117,40 @@ const unsubscribeRecoverObjectMappings = (subscriptionId: string) => {
   }
 }
 
+// Limits the maximum piece index to the already archived segments
+// avoiding to distribute pieces that are not yet archived
+const getUpperLimit = (pieceIndex: number) => {
+  const segmentIndex = segmentUseCase.getSegmentByPieceIndex(pieceIndex)
+  if (segmentIndex >= state.lastRealtimeSegmentIndex) {
+    const UPPER_LIMIT_RANGE_INDEX = 1
+    return segmentUseCase.getPieceIndexRangeBySegment(
+      state.lastRealtimeSegmentIndex,
+    )[UPPER_LIMIT_RANGE_INDEX]
+  }
+  return pieceIndex
+}
+
 const emitRecoverObjectMappings = async () => {
   const recovering = Array.from(
     state.recoverObjectMappingsSubscriptions.entries(),
   )
 
   const promises = recovering.map(
     async ([subscriptionId, { connection, pieceIndex, step }]) => {
-      const result = await objectMappingUseCase.getObjectByPieceIndexAndStep(
+      const upperLimit = getUpperLimit(pieceIndex + step)
```

Comment:
Additionally, for peers subscribed to recovering object mappings we assure that the requested range does not included pieces whose segments are not fully archived.